src/llvm-julia-licm.cpp

```diff
@@ -112,6 +117,44 @@ struct JuliaLICMPass : public LoopPass, public JuliaPassContext {
                         CallInst::Create(call, {}, exit_pts[i]);
                     }
                 }
+                else if (callee == write_barrier_func) {
+                    bool valid = true;
+                    for (std::size_t i = 0; i < call->getNumArgOperands(); i++) {
+                        if (!L->makeLoopInvariant(call->getArgOperand(i), changed)) {
+                            valid = false;
+                            break;
+                        }
+                    }
+                    if (valid) {
+                        call->moveBefore(preheader->getTerminator());
+                        changed = true;
+                    }
+                }
+                else if (callee == alloc_obj_func) {
+                    jl_alloc::AllocUseInfo use_info;
+                    jl_alloc::CheckInst::Stack check_stack;
+                    jl_alloc::EscapeAnalysisRequiredArgs required{use_info, check_stack, *this, DL};
+                    jl_alloc::runEscapeAnalysis(call, required, jl_alloc::EscapeAnalysisOptionalArgs().with_valid_set(&L->getBlocksSet()));
+                    if (use_info.escaped || use_info.addrescaped) {
```

Comment:
Just an observation: this is depending on more than just escape analysis results, but also is depending on the absence of loop-carried dependencies. Currently, `runEscapeAnalysis` has no handling implemented for PhiNode, so it looks like it is correct, but if the escape-analysis ever gains the ability to look at the function CFG, we need to fix this usage.